### DISCUSSION TITLE
Install packages before files

### DISCUSSION THREAD

**Discussion by MenacingPerson** [CyberShadow/aconfmgr#140] [Q&A]
The other way around seems to be the default, I wonder why? It breaks pacman on a fresh arch install because i have [ALHP](https://git.harting.dev/ALHP/ALHP.GO).

**Comment by CyberShadow** [comment]
aconfmgr installs certain files before doing anything else, as these files affect the rest of the process. For example, pacman configuration files are installed before running pacman, because these may be required to install packages in third-party repositories. Another example is `/etc/passwd`, which is needed to pre-configure UNIX users to be able to then `chown` files to them, and avoid installed packages from creating these with other UIDs.

Here is the list:
https://github.com/CyberShadow/aconfmgr/blob/eb220118a22488de7c3db2c5a9576b024b36fca5/src/common.bash#L71-L77

Depending on your intent, you could modify this array in your aconfmgr configuration, either to add additional files to be available to pacman early, or to exclude the pacman configuration file.

**Reply by MenacingPerson** [reply]
How about allowing people to "force" certain packages to install before these files?

**Reply by CyberShadow** [reply]
I think you could achieve that effect by adding `sudo pacman -S --needed PACKAGE` directly to the aconfmgr configuration. See also `AconfNeedProgram`.

**Reply by MenacingPerson** [reply]
> I think you could achieve that effect by adding `sudo pacman -S --needed PACKAGE` directly to the aconfmgr configuration

What do you mean by that?

Like, instead of AddPackage (or with it because then aconfmgr would complain), you have that in there?

> AconfNeedProgram

Not documented anywhere, would that be the same as the above one?

**Reply by CyberShadow** [reply]
> What do you mean by that?
> 
> Like, instead of AddPackage (or with it because then aconfmgr would complain), you have that in there?

Yes, so, do this one thing imperatively rather than declaratively.

> Not documented anywhere, would that be the same as the above one?

Yes, not documented on its own at the moment, but it's mentioned in an example in the README.

**Reply by MenacingPerson** [reply]
AconfNeedProgram dies with this error:

![image](https://user-images.githubusercontent.com/59611020/158026519-b8286cd6-40b4-4d00-be82-6cc8bb184575.png)

Here's the corresponding code: https://github.com/Samisafool/dotfiles/blob/main/.config/aconfmgr/30-nongui.sh#L70

**Reply by CyberShadow** [reply]
What AconfNeedProgram does is that it installs a package if the specified binary is not in PATH. Here is the implementation:
https://github.com/CyberShadow/aconfmgr/blob/eb220118a22488de7c3db2c5a9576b024b36fca5/src/common.bash#L1283-L1302

**Reply by MenacingPerson** [reply]
Ohhhh, I understand now...

I suppose I'll just make $1 be `Samisafool` or something...

**Reply by MenacingPerson** [reply]
Works now! (Except for #33, but mostly unrelated)
